kernel/mellanox_ethernet/src/command_queue.rs

```diff
@@ -0,0 +1,740 @@
+                cmdq_entry.set_input_length_in_bytes(0x10 + pages_pa.len() as u32 * SIZE_PADDR_IN_BYTES as u32); 
+                cmdq_entry.set_output_length_in_bytes(16);
+                cmdq_entry.set_input_inline_data(
+                    opcode, 
+                    opmod, 
+                    None, 
+                    Some(pages_pa.len() as u32)
+                );
+
+                self.init_manage_pages_input_mailbox_buffers(entry_num, pages_pa)?;
+                self.set_mailbox_pointer_in_cmd_entry(&mut cmdq_entry, entry_num, MailboxType::Input);
+            }
+            CommandOpcode::AllocUar => {
+                cmdq_entry.set_input_length_in_bytes(8);
+                cmdq_entry.set_output_length_in_bytes(12);
+                cmdq_entry.set_input_inline_data(opcode, opmod, None, None);
+            },
+            CommandOpcode::QueryVportState => { // only accesses your own vport
+                cmdq_entry.set_input_length_in_bytes(12);
+                cmdq_entry.set_output_length_in_bytes(16);
+                cmdq_entry.set_input_inline_data(opcode, opmod, None, None);
+            },
+            CommandOpcode::AllocPd => { 
+                cmdq_entry.set_input_length_in_bytes(8);
+                cmdq_entry.set_output_length_in_bytes(12);
+                cmdq_entry.set_input_inline_data(opcode, opmod, None, None);
+            },
+            CommandOpcode::AllocTransportDomain => { 
+                cmdq_entry.set_input_length_in_bytes(8);
+                cmdq_entry.set_output_length_in_bytes(12);
+                cmdq_entry.set_input_inline_data(opcode, opmod, None, None);
+            },
+            CommandOpcode::QuerySpecialContexts => { 
+                cmdq_entry.set_input_length_in_bytes(8);
+                cmdq_entry.set_output_length_in_bytes(12);
+                cmdq_entry.set_input_inline_data(opcode, opmod, None, None);
+            },
+            _=> {
+                error!("unimplemented opcode");
+                return Err("Unimplemented opcode!");
+            }
+        }        
+
+        // write command to actual place in memory
+        core::mem::swap(&mut cmdq_entry, &mut self.entries[entry_num]);
+        
+        // update token to be used in the next command
+        self.token = self.token.wrapping_add(1);
+
+        // claim the command entry as in use
+        self.available_entries[entry_num] = false;
+
+        Ok(entry_num)
+    }
+
+    /// Sets the mailbox pointer in a command entry with the physical address of the first mailbox.
+    fn set_mailbox_pointer_in_cmd_entry(&mut self, cmdq_entry: &mut CommandQueueEntry, entry_num: usize, mailbox_type: MailboxType) {
+        let mailbox_ptr = if mailbox_type == MailboxType::Input {
+            self.mailbox_buffers_input[entry_num][0].1.value()
+        } else {
+            self.mailbox_buffers_output[entry_num][0].1.value()
```

Comment:
in coordination with my other comments about avoiding indexing and avoiding passing around `entry_num`, this one has a double index and should definitely be restructured. 

Here's, I think it'd be better to make this an associated function (rather than a method that takes `&mut self`) and pass a mutable reference to the `mailbox_buffers_*` and then modify that parameter directly. 

you may wish to break it up into two functions, one for `MailboxType::Input` and one for `Output`